File: src/php/t5/s1.php
<?php

declare(strict_types=1);

function k_complement(int $k, array $a): int {
    $amountComplimentary = 0;

    /**
     * Hold amount of occurrence of a single number inside starting array '$a'
     * Structure is:
     *  [
     *      [
     *          key: number
     *          value: amount occurrence in original array '$a'
     *      ],
     *      [
     *          ...
     *      ],
     *  ]
     */
    $occurrencesNumbers = [];

    foreach ($a as $number) {
        if (!array_key_exists($number, $occurrencesNumbers)) {
            $occurrencesNumbers[$number] = 1;
        } else {
            $occurrencesNumbers[$number]++;
        }
    }

    $numbers = array_keys($occurrencesNumbers);

    foreach ($numbers as $number) {
        // Building the complementary pair by subtracting one number from the target result
        // and checking if the result (other pair part) exists
        $result = $k - $number;
        if (in_array($result, $numbers)) {
            $amountComplimentary += $occurrencesNumbers[$number] * $occurrencesNumbers[$result];
        }
    }

    return $amountComplimentary;
}

File: src/go/t5/s1.go
package main

func Kcomplement(k int, numbers []int) int {
	pairs := 0
	occurrencesNumbers := make(map[int]int)

	for _, number := range numbers {
		if _, ok := occurrencesNumbers[number]; ok {
			occurrencesNumbers[number]++
		} else {
			occurrencesNumbers[number] = 1
		}
	}

	for number, occurrence := range occurrencesNumbers {
		result := k - number
		if _, ok := occurrencesNumbers[number]; ok {
			pairs += occurrence * occurrencesNumbers[result]
		}
	}

	return pairs
}
